sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,667 @@
+    desc.vendor_id = USB_CONFIG_VID;
+    desc.product_id = USB_CONFIG_PID;
+    desc.manufacturer_idx = usbus->manuf.idx;
+    desc.product_idx = usbus->product.idx;
+    desc.num_configurations = 1;
+    usbus_put_bytes(usbus, (uint8_t*)&desc, sizeof(usb_descriptor_device_t));
+    usbus_ep0_ready(usbus);
+}
+
+static void _req_config(usbus_t *usbus)
+{
+    mutex_lock(&usbus->lock);
+    usbus_hdrs_fmt_conf(usbus);
+    mutex_unlock(&usbus->lock);
+    usbus_ep0_ready(usbus);
+}
+
+static void _req_dev_qualifier(usbus_t *usbus)
+{
+    usb_speed_t speed = USB_SPEED_LOW;
+    usbus->dev->driver->get(usbus->dev, USBOPT_MAX_SPEED, &speed, sizeof(usb_speed_t));
+    if (speed == USB_SPEED_HIGH) {
+        /* TODO: implement device qualifier support */
+    }
+    /* Signal stall to indicate unsupported (USB 2.0 spec 9.6.2 */
+    static const usbopt_enable_t enable = USBOPT_ENABLE;
+
+    usbus->in->driver->set(usbus->in, USBOPT_EP_STALL, &enable, sizeof(usbopt_enable_t));
+}
+
+static void _req_descriptor(usbus_t *usbus, usb_setup_t *pkt)
+{
+    uint8_t type = pkt->value >> 8;
+    uint8_t idx = (uint8_t)pkt->value;
+    switch (type) {
+        case 0x1:
+            _req_dev(usbus);
+            break;
+        case 0x2:
+            _req_config(usbus);
+            break;
+        case 0x03:
+            _req_str(usbus, idx);
+            break;
+        case 0x06:
+            _req_dev_qualifier(usbus);
+            break;
+        default:
+            break;
+    }
+}
+
+void recv_dev_setup(usbus_t *usbus, usbdev_ep_t *ep, usb_setup_t *pkt)
+{
+    (void)ep;
+    if (pkt->type & 0x80) {
+        switch (pkt->request) {
+            case 0x00:
+                _req_status(usbus);
+                break;
+            case 0x06:
+                _req_descriptor(usbus, pkt);
+                break;
+            default:
+                break;
+        }
+    }
+    else{
+        switch (pkt->request) {
+            case 0x05:
+                usbus->addr = (uint8_t)pkt->value;
+                break;
+            case 0x09:
+                _activate_endpoints(usbus);
+                break;
+            default:
+                break;
+        }
+        /* Signal zlp */
+        usbus->in->driver->ready(usbus->in, 0);
+    }
+}
+
+void recv_interface_setup(usbus_t *usbus, usbdev_ep_t *ep, usb_setup_t *pkt)
+{
+    uint16_t destination = pkt->index & 0x0f;
+    (void)ep;
+    /* Find interface handler */
+    mutex_lock(&usbus->lock);
+    for (usbus_interface_t *iface = usbus->iface; iface; iface = iface->next) {
+        if (destination == iface->idx) {
+            iface->handler->driver->event_handler(usbus, iface->handler, USBUS_MSG_TYPE_SETUP_RQ, pkt);
+        }
+    }
+    mutex_unlock(&usbus->lock);
+}
+
+
+static inline size_t usbus_pkt_maxlen(usbus_t *usbus, usb_setup_t *pkt)
+{
+    return pkt->length > usbus->in->len ? usbus->in->len : pkt->length;
+}
+
+void recv_setup(usbus_t *usbus, usbdev_ep_t *ep)
```

Comment:
You probably know better than I, do we have a function naming convention? I mean, since C has a single name space, there could easily arise a name conflict with functions named like this.
If this becomes a `static` function then there is no problem, of course.

Reply:
Should probably also be static

Reply:
> do we have a function naming convention?

Yes, public (non-static) functions should be called ```module_function()``` (e.g., ```usbus_recv_setup()```). static functions should start with underline (```static void _recv_setup(void)```).

Reply:
marked almost all these functions as static, please let me know if I missed one